### PR TITLE
add confirmation aeas for aw2

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR adds the confirmation AEAs for aw2. It also modifies the generic buyer.

## Fixes

-

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: packages/fetchai/skills/confirmation_aw2/registration_db.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            "CREATE TABLE IF NOT EXISTS trade_table (address TEXT, first_trade timestamp, "
+            "second_trade timestamp)"
+        )
+
+    def set_trade(
+        self, address: str, timestamp: datetime.datetime, data: Dict[str, str],
+    ):
+        """Record a registration."""
+        is_second = False  # todo, query from db
+        # todo: also save data
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/confirmation_aw2/registration_db.py:70]: The idea; we need to save the trade including its data

File: packages/fetchai/skills/confirmation_aw2/registration_db.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            )
+        variables = (
+            address,
+            timestamp,
+        )
+        self._execute_single_sql(command, variables)
+
+    def is_registered(self, address: str) -> bool:
+        """Check if an address is registered."""
+        command = "SELECT * FROM registered_table WHERE address=?"
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/confirmation_aw2/registration_db.py:87]: The registered_table is taken from confirmation_aea_aw1

File: packages/fetchai/skills/confirmation_aw2/registration_db.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    def is_allowed_to_trade(self, address: str, mininum_hours_between_txs: int) -> bool:
+        """Check if an address is registered."""
+        command = "SELECT * FROM trade_table WHERE address=?"
+        variables = (address,)
+        result = self._execute_single_sql(command, variables)
+        # todo: check :
+        #  - if no timestamp present: return true
+        #  - if first timestamp is present: return true iff the trade is 4 hours in past
+        #  - if second timestamp is present: return false
+        return result is None
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/confirmation_aw2/registration_db.py:101]: see todo comments

File: packages/fetchai/skills/confirmation_aw2/README.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+## Behaviours
+
+* search: searches for data selling service on SOEF
+
+## Handlers
+
+* fipa: handles fipa messages for negotiation
+* ledger_api: handles ledger_api messages for payment
+* oef_search: handles oef_search messages to manage the sellers it finds
+* signing: handles signing messages for transaction signing by the decision maker
```
Comment (marcofavorito) [inline packages/fetchai/skills/confirmation_aw2/README.md:19]:
```suggestion
* `fipa`: handles fipa messages for negotiation
* `ledger_api`: handles ledger_api messages for payment
* `oef_search`: handles oef_search messages to manage the sellers it finds
* `signing`: handles signing messages for transaction signing by the decision maker
```

File: packages/fetchai/skills/confirmation_aw2/README.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+## Description
+
+This skill purchases information from other agents as specified in its configuration. It acts as the confirmation AEA in Agent World 2.
+
+This skill searches for an agent in a vicinity on the SOEF that sells the data this skill is configured to buy. Once it found some agents which match Agent World 2 criteria, it requests this data, negotiates the price, pays the proposed amount if agreement is reached, and receives the data bought.
+
+
+## Behaviours
+
+* search: searches for data selling service on SOEF
```
Comment (marcofavorito) [inline packages/fetchai/skills/confirmation_aw2/README.md:12]:
```suggestion
* `search`: searches for data selling service on SOEF
```

File: packages/fetchai/skills/confirmation_aw2/README.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+* fipa: handles fipa messages for negotiation
+* ledger_api: handles ledger_api messages for payment
+* oef_search: handles oef_search messages to manage the sellers it finds
+* signing: handles signing messages for transaction signing by the decision maker
+
+
+## Models
+
+* strategy: allows the configuration of the purchasing. In particular, location and search_radius together determine the vicinity where the service is searched for, search_query specifies the query, and the remaining configuration specifies the terms of trade. Also, rules regarding Agent World 2 are enforced.
```
Comment (marcofavorito) [inline packages/fetchai/skills/confirmation_aw2/README.md:24]:
```suggestion
* `strategy`: allows the configuration of the purchasing. In particular, location and search_radius together determine the vicinity where the service is searched for, search_query specifies the query, and the remaining configuration specifies the terms of trade. Also, rules regarding Agent World 2 are enforced.
```

File: packages/fetchai/skills/confirmation_aw2/registration_db.py [hunk 7]
```diff
@@ -65,26 +65,28 @@ def _initialise_backend(self) -> None:
...
+            variables: Tuple[
+                str, datetime.datetime, Optional[datetime.datetime], str, Optional[str]
+            ] = (address, timestamp, None, json.dumps(data), None)
+        else:
+            _, first_trade, second_trade, first_info, _ = record
+            is_second = first_trade is not None and second_trade is None
+            is_more_than_two = first_trade is not None and second_trade is not None
+            if is_more_than_two or not is_second:
+                return
             command = "INSERT INTO trade_table(address, first_trade, second_trade, first_info, second_info) values(?, ?, ?, ?, ?)"
```
Comment (marcofavorito) [inline packages/fetchai/skills/confirmation_aw2/registration_db.py:82]: @DavidMinarsch this should be `INSERT or REPLACE into ...`

Comment (marcofavorito) [reply packages/fetchai/skills/confirmation_aw2/registration_db.py:82]: notice also that `address` now is a primary key

Comment (DavidMinarsch) [reply packages/fetchai/skills/confirmation_aw2/registration_db.py:82]: Yes!

File: packages/fetchai/skills/confirmation_aw2/registration_db.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            # no record on trade: go ahead
+            return True
+        first_trade: Optional[datetime.datetime] = record[1]
+        second_trade: Optional[datetime.datetime] = record[2]
+        first_trade_present: bool = first_trade is not None
+        second_trade_present: bool = second_trade is not None
+        if not first_trade_present and not second_trade_present:
+            # all trades empty: go ahead
+            return True
+        if first_trade is not None and not second_trade_present:
```
Comment (marcofavorito) [inline packages/fetchai/skills/confirmation_aw2/registration_db.py:118]: minor
```suggestion
        if not first_trade_present and not second_trade_present:
```

**Review by marcofavorito — APPROVED** [review]
LGTM

**Pull request merged**